### ISSUE TITLE
3.2.0: build fails with gcc 11.0.1

### ISSUE THREAD

**Issue by kloczek** [fish-shell/fish-shell#7764]
```
[  3%] Building CXX object CMakeFiles/fishlib.dir/src/fd_monitor.cpp.o
/usr/bin/g++ -DBINDIR=L\"/usr/bin\" -DCMAKE_BINARY_DIR=\"/home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0/x86_64-redhat-linux-gnu\" -DCMAKE_SOURCE_DIR=\"/home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0\" -DDATADIR=L\"/usr/share\" -DDOCDIR=L\"/usr/share/doc/fish\" -DLOCALEDIR=\"/usr/share/locale\" -DPREFIX=L\"/usr\" -DSYSCONFDIR=L\"/etc\" -D_REENTRANT -D_UNICODE=1 -I/home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0/x86_64-redhat-linux-gnu -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -Wall -Wextra -Wno-comment -Wno-address  -Wredundant-move -O2 -g  -fno-exceptions -pthread -std=gnu++11 -o CMakeFiles/fishlib.dir/src/fd_monitor.cpp.o -c /home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0/src/fd_monitor.cpp
/home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0/src/fd_monitor.cpp: In destructor ‘fd_monitor_t::~fd_monitor_t()’:
/home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0/src/fd_monitor.cpp:25:27: error: ‘sleep_for’ is not a member of ‘std::this_thread’
   25 |         std::this_thread::sleep_for(std::chrono::milliseconds(5));
      |                           ^~~~~~~~~
make[3]: *** [CMakeFiles/fishlib.dir/build.make:725: CMakeFiles/fishlib.dir/src/fd_monitor.cpp.o] Error 1
```
Looks like 3.2.0 now uses exceptions and rtti so now not longer is possible to build fish binaries with `-fno-rtti -fno-exceptions` to reduce code size by about 60KB of each binary ..

**Comment by faho** [comment]
>Looks like 3.2.0 now uses exceptions and rtti

We [already disable exceptions ourselves](https://github.com/fish-shell/fish-shell/blob/master/CMakeLists.txt#L52)

**Comment by faho** [comment]
Given that the error is that "std::this_thread" doesn't have a "sleep_for" function, this looks like your standard library is broken, or you're trying to compile fish without C++11.

**Comment by faho** [comment]
Ah, okay, disabling rtti is indeed not possible anymore - we have a few `dynamic_cast`s now.

Anyway, this issue looks like a problem in your compiler/stdlib, the sleep_for is entirely ordinary and should just work.

**Comment by kloczek** [comment]
```
[  2%] Building CXX object CMakeFiles/fishlib.dir/src/exec.cpp.o
/usr/bin/g++ -DBINDIR=L\"/usr/bin\" -DCMAKE_BINARY_DIR=\"/home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0/x86_64-redhat-linux-gnu\" -DCMAKE_SOURCE_DIR=\"/home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0\" -DDATADIR=L\"/usr/share\" -DDOCDIR=L\"/usr/share/doc/fish\" -DLOCALEDIR=\"/usr/share/locale\" -DPREFIX=L\"/usr\" -DSYSCONFDIR=L\"/etc\" -D_REENTRANT -D_UNICODE=1 -I/home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0/x86_64-redhat-linux-gnu -O2 -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fdata-sections -ffunction-sections -flto=auto -flto-partition=none -fno-rtti -fno-exceptions -Wall -Wextra -Wno-comment -Wno-address  -Wredundant-move -O2 -g  -fno-exceptions -pthread -std=gnu++11 -o CMakeFiles/fishlib.dir/src/exec.cpp.o -c /home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0/src/exec.cpp
/home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0/src/exec.cpp: In function ‘std::unique_ptr<output_stream_t> create_output_stream_for_builtin(int, const io_chain_t&, bool)’:
/home/tkloczko/rpmbuild/BUILD/fish-shell-3.2.0/src/exec.cpp:448:17: error: ‘dynamic_cast’ not permitted with ‘-fno-rtti’
  448 |                 dynamic_cast<const io_bufferfill_t *>(io.get())->buffer();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/fishlib.dir/build.make:686: CMakeFiles/fishlib.dir/src/exec.cpp.o] Error 1
```

**Comment by kloczek** [comment]
> We already disable exceptions ourselves, so that can't be it

Myfault and you are right :)
I've been looking on wrong terminal with other package `-fno-exceptions` build error :)

**Comment by faho** [comment]
>error: ‘dynamic_cast’ not permitted with ‘-fno-rtti’

Indeed, this means no-rtti isn't possible anymore, like I said.

Now I'm a bit confused as to what happened to the first error? Did you solve that? This new one is just no-rtti not being supported, so you need to stop doing that.

**Comment by faho** [comment]
Alright, it turns out that was our *only* dynamic_cast, so I just replaced it with an oldschool C-cast and built with -fno-rtti to confirm (with gcc 10.2).

Whatever happened to have sleep_for break I don't know, but I'm confident it's an issue in your toolchain, not fish, so I'm closing this. If you have any reason to suspect it's a problem with fish, we'll reopen.

**Comment by kloczek** [comment]
Again: this is issue gcc 11.0.1 specyfic.
Please have look that issue one more time.

And about rtti. Cost of using that rtti just one time is that now fish binary cannot be ~60KB smaller.

**Comment by faho** [comment]
>And about rtti. Cost of using that rtti just one time is that now fish binary cannot be ~60KB smaller.

It's 60 *kilo*bytes. That's basically nothing. I've tested it, it doesn't make a difference in any benchmarks, it's not going to fill your harddrive. Even in relative terms it's 3% of fish's size.

Anyway, that bit is fixed now, I removed the dynamic_cast. Disable rtti to your heart's content for now. However: If we ever find an actual use for it, we will use it.

> Please have look that issue one more time.

The only possible thing I can see is that fd_monitor.cpp doesn't explicitly `include <thread>` - only transitively via iothread. Try adding that and see if it helps.
